pkg/cryptutil/kek.go

```diff
@@ -0,0 +1,165 @@
+package cryptutil
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+
+	"github.com/google/uuid"
+	"golang.org/x/crypto/curve25519"
+	"golang.org/x/crypto/nacl/box"
+)
+
+// A KeyEncryptionKey (KEK) is used to implement *envelope encryption*, similar to how data is stored at rest with
+// AWS or Google Cloud:
+//
+//   - AWS: https://docs.aws.amazon.com/kms/latest/developerguide/concepts.html#enveloping
+//   - Google Cloud: https://cloud.google.com/kms/docs/envelope-encryption
+//
+// Data is encrypted with a data encryption key (DEK) and that key is stored next to the data encrypted with the KEK.
+// Finally the KEK id is also stored with the data.
+//
+// To decrypt the data you first retrieve the KEK, second decrypt the DEK, and finally decrypt the data using the DEK.
+//
+//   - Our KEKs are asymmetric Curve25519 keys. We use the *public* key to encrypt the DEK so only the *private* key can
+//     decrypt it.
+//   - Our DEKs are symmetric XChaCha20Poly1305 keys.
+//
+type KeyEncryptionKey interface {
+	ID() string
+	KeyBytes() []byte
+
+	isKeyEncryptionKey()
+}
+
+// KeyEncryptionKeySize is the size of a key encryption key.
+const KeyEncryptionKeySize = curve25519.ScalarSize
+
+// PrivateKeyEncryptionKey is a Curve25519 asymmetric private encryption key used to decrypt data encryption keys.
+type PrivateKeyEncryptionKey struct {
+	id   string
+	data [KeyEncryptionKeySize]byte
+}
+
+func (*PrivateKeyEncryptionKey) isKeyEncryptionKey() {}
+
+// NewPrivateKeyEncryptionKey creates a new encryption key from existing bytes.
+func NewPrivateKeyEncryptionKey(id string, raw []byte) (*PrivateKeyEncryptionKey, error) {
+	if len(raw) != KeyEncryptionKeySize {
+		return nil, fmt.Errorf("cryptutil: invalid key encryption key, expected %d bytes, got %d",
+			KeyEncryptionKeySize, len(raw))
+	}
+	kek := new(PrivateKeyEncryptionKey)
+	kek.id = id
+	copy(kek.data[:], raw)
+	return kek, nil
+}
+
+// GenerateKeyEncryptionKey generates a new random key encryption key.
+func GenerateKeyEncryptionKey() (*PrivateKeyEncryptionKey, error) {
+	raw := make([]byte, KeyEncryptionKeySize)
+	_, err := io.ReadFull(rand.Reader, raw)
+	if err != nil {
+		return nil, fmt.Errorf("cryptutil: error generating random data encryption key: %w", err)
+	}
+	return NewPrivateKeyEncryptionKey(uuid.New().String(), raw)
```

Comment:
Thoughts on using the sha256-hmac of the key itself for this value to:
1. Limit depedencies
2. If, for whatever reason, you have just the private key, you can find out the associated key-id

Reply:
I used our existing `Hash` function and base58.